### PR TITLE
Global SNR VDrop Pruning

### PR DESCRIPTION
**Pull request by alexcuozzo-1**
https://ntaresearch.atlassian.net/browse/RES-2264

### CODE REVIEW

**Comment by alexcuozzo-1** [comment]
In the global SNR pruning mixin, I referenced the model's VDropCentralData as vdrop_data = self.model.vdrop_central_data, but in the original layer-wise version it is referred to as vdrop_data = self.model.module.vdrop_central_data. This would require the model to assign its vdrop to vdrop_central_data. I would like to keep the convention consistent between the two mixins but wasn't sure which to use - any thoughts?

File: nupic/research/frameworks/vernon/mixins/prune_low_snr.py [hunk 1]
```diff
@@ -116,9 +109,90 @@ def run_epoch(self):
...
+            num_weights = math.floor(z_logalpha.numel() * target_density)
+            on_indices = z_logalpha.topk(num_weights, largest=False)[1]
+
+            z_mask.zero_()
+            z_mask[on_indices] = 1
+            z_mu.data *= z_mask
+            with torch.no_grad():
+                z_logvar[~z_mask.bool()] = vdrop_data.pruned_logvar_sentinel
+
+            if not self.logger.disabled:
```
Comment (mrcslws) [inline nupic/research/frameworks/vernon/mixins/prune_low_snr.py:156]: This check isn't necessary. It's already built into `logger.info`. We only need to do this type of check if we need to compute something expensive before logging.

**Review by mrcslws — COMMENTED** [review]
RE your comment, the issue is DistributedDataParallel, which wraps the model. You're not using it, but we use it elsewhere. We might just need the code to support both...

```python
model = self.model
if hasattr(model, "module"):
    model = model.module

vdrop_data = model.vdrop_central_data
```

(though this is admittedly annoying)

File: nupic/research/frameworks/vernon/mixins/prune_low_snr.py [hunk 2]
```diff
@@ -116,9 +109,90 @@ def run_epoch(self):
...
+    """
+    Prunes weights by signal-to-noise ratio across the entire model. Requires
+    variational dropout modules. For this mixin to work, the model must have a
+    VDropCentralData module stored in the model as self.vdrop_central_data.
+    """
+
+    def setup_experiment(self, config):
+        """
+        :param config:
+            - prune_schedule: A list of (epoch, sparsity) pairs.
```
[the remark this answers is not in the document]
Comment (alexcuozzo-1) [reply nupic/research/frameworks/vernon/mixins/prune_low_snr.py:128]: yes, changed

**Review by mrcslws — APPROVED** [review]

**Pull request merged**